_posts/2021-08-15-interviews.markdown

```diff
@@ -0,0 +1,155 @@
+
+~~ Tasks that will execute ~~
+push_values - populate the list - {'elements': [1, 2, 3, 4]}
+print_list - print the list - {}
+reverse_list - reverse the list - {}
+print_list - print the reversed list - {}
+INFO:root:~~~
+INFO:root:Populating the list with: [1, 2, 3, 4]
+INFO:root:~~~
+the list is
+4, 3, 2, 1
+INFO:root:~~~
+INFO:root:Reversing the list
+INFO:root:~~~
+the list is
+1, 2, 3, 4
+```
+
+We did it! We reversed a linked list. Want to add [some other task for a `LinkedList`](https://www.geeksforgeeks.org/top-20-linked-list-interview-question/)? No problem, we just have to:
+
+- Define a new `action` for the yaml config.
+- Extend `Task` with a new subclass and implement `execute`.
+- Write some unit & integration tests, which are glaringly absent from this post!
+- Add a line to `task_factory` to produce the new class for the appropriate value of `action`.
+- Document it and communicate it to the users of your platform.
+
+## The twist
+
+Most coding interviews have a secret second part. If a candidate solves the problem with lots of time to spare, we can ask an add-on question. Is this fair? Probably not. If a candidate can pass the interview by solving just the first part, they shouldn't be able to fail it by attempting the second. Hopefully by solving the second part of the question they considered for a higher title, and not penalized for getting stuck on it.
+
+In this case, we could make a more complex task execution graph. The execution graph looks like this in v1.
+
+```
+push_values -> print_list -> reverse_list -> print_list
+```
+
+The first `print_list` task and the `reverse_list` tasks are really operating on the same input (`print_list` is really just a side effect). We could allow each task to specify its parent and re-draw the graph to look like this.
+
+```
+               / -> print_values
+              /
+push_values -
+              \
+               \ -> reverse_list -> print_values
+```
+
+So we can change the yaml config format a bit to include an `id` for each task and a `parent` indicating the ID of upstream task. The new config will look like this. It's a breaking change, so we've moved to `apiVersion: 2`.
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=pipeline2.yaml" type="text/javascript"></script>
+
+Now we can implement a `Parser2` that's a little more advanced. Rather than a `List[Task]` we'll create a dictionary from `task_id -> List[downstream Tasks]`, and step through all of the tasks. Here's how `run` is implemented now (see the link at the bottom for the full code):
+
+<script src="https://gist.github.com/35a63e7631b60641fc0342f31461b80d.js?file=parser2.py" type="text/javascript"></script>
+
+I made a _graph_ of parent and child tasks, and then step through starting at the head and hitting every child, passing the output of the parent to each of its children tasks. **It's a depth-first graph traversal** just hiding there right in front of us!
+
+## Wrap-up
```

Comment:
as mentioned above I think you could reiterate here the various reasons why you think this approach is better, such as

- the task is more life-like
- lots of interesting stuff to talk about in how they design the code, interface, etc
- lots of ways to evolve the question to do other interesting things they'd also have to tackle in the real world some day, like the graph stuff